cjwstate/modules/message_catalogs.py

```diff
@@ -0,0 +1,116 @@
+from cjworkbench.i18n import supported_locales, default_locale
+from cjworkbench.i18n.catalogs.util import (
+    read_po_catalog,
+    write_po_catalog,
+    remove_strings,
+    find_fuzzy_messages,
+    fill_catalog,
+    mark_fuzzy,
+    MessageUID,
+)
+from babel.messages.catalog import Catalog
+from cjwstate.modules.module_loader import ModuleFiles, ModuleSpec
+import os.path
+from typing import FrozenSet
+
+
+def update_module_catalogs(spec: ModuleSpec):
+    source_catalog = _build_source_catalog(spec)
+    po_path = _po_path(directory, default_locale, spec.id_name)
+    try:
+        old_source_catalog = read_po_catalog(po_path)
+    except FileNotFoundError:
+        old_source_catalog = Catalog(default_locale)
+    fuzzy = find_fuzzy_messages(
+        old_catalog=old_source_catalog, new_catalog=source_catalog
+    )
+    write_po_catalog(po_path, source_catalog)
+
+    for locale_id in supported_locales:
+        po_path = _po_path(directory, locale_id, spec.id_name)
+        try:
+            old_catalog = read_po_catalog(po_path)
+        except FileNotFoundError:
+            old_catalog = Catalog(locale_id)
+        write_po_catalog(
+            po_path, _merge_nonsource_catalog(old_catalog, source_catalog, fuzzy)
+        )
+
+
+def _po_path(basepath: os.path, locale_id: str, module_id_name: str) -> os.path:
+    return os.path.join(basepath, "locale", locale_id, f"{module_id_name}.po")
+
+
+def _build_source_catalog(spec: ModuleSpec) -> Catalog:
+    messages = _find_spec_messages(spec)
+    # TODO: also find messages in module code
+    source_catalog = Catalog(default_locale)
+    for message_id, source_string in messages.items():
+        if source_string is not None:
+            source_catalog.add(message_id, string=source_string)
+
+    return source_catalog
+
+
+def _merge_nonsource_catalog(
+    old_catalog: Catalog, source_catalog: Catalog, fuzzy: FrozenSet[MessageUID]
+) -> Catalog:
+    catalog = Catalog(locale_id)
+    fill_catalog(catalog, source_catalog, old_catalog)
+    mark_fuzzy(catalog, fuzzy, old_catalog)
+    return catalog
+
+
+def _find_spec_messages(spec: ModuleSpec) -> Catalog:
+    prefix = f"modules.{spec.id_name}._spec"
```

Comment:
Let's nix the `prefix` var and just begin every key with `_spec` (which has one fewer character than `prefix`).

I don't think we want to start every i18n key with `modules.{id_name}.` Is there a good reason to have that longer prefix?